File: examples/src/example_2_hardcoded_obstacles_gazebo.cpp
#include "example_common.hpp"

#include "sclerp/collision/collision.hpp"
#include "sclerp/collision/motion_plan_collision.hpp"
#include "sclerp/collision/robot_link_meshes.hpp"
#include "sclerp/core/kinematics/kinematics_solver.hpp"
#include "sclerp/core/planning/motion_plan.hpp"
#include "sclerp/gazebo/world_registry.hpp"
#include "sclerp/gazebo/world_sdf.hpp"
#include "sclerp/trajectory/interpolator.hpp"
#include "sclerp/urdf/load_manipulator.hpp"

#include <Eigen/Dense>
#include <Eigen/Geometry>

#include <cmath>
#include <filesystem>
#include <iostream>
#include <stdexcept>
#include <string>

namespace {

static void printUsage(const char* exe) {
  std::cout << "Example 2: plan with hardcoded obstacles (collision module), time-parameterize, export Gazebo world + CSV.\n\n"
            << "Usage:\n"
            << "  " << exe << " [--urdf <path> --stl-dir <dir> --base-link <name> --tip-link <name>] [options]\n\n"
            << "If robot args are omitted, uses bundled Diana7 assets.\n\n";
  sclerp::examples::printCommonCli(std::cout);
}

static void registerObstaclesOrThrow(sclerp::gazebo::WorldRegistry* reg,
                                     const std::vector<std::shared_ptr<sclerp::collision::FclObject>>& obstacles) {
  if (!reg) throw std::runtime_error("null WorldRegistry");
  if (obstacles.empty()) return;

  for (std::size_t i = 0; i < obstacles.size(); ++i) {
    std::string name;
    if (i == 0) name = "obstacle_box";
    else if (i == 1) name = "obstacle_cylinder";
    else name = "obstacle_" + std::to_string(i);
    const auto st = reg->registerObstacle(obstacles[i], name);
    if (!sclerp::core::ok(st)) {
      throw std::runtime_error("WorldRegistry::registerObstacle failed for " + name);
    }
  }
}

}  // namespace

int main(int argc, char** argv) {
  try {
    const auto parsed = sclerp::examples::parseArgs(argc, argv);
    if (parsed.flags.find("--help") != parsed.flags.end()) {
      printUsage(argv[0]);
      return 0;
    }

    const auto args = sclerp::examples::parseCommonArgsOrThrow(parsed);
    sclerp::examples::ensureOutputDirOrThrow(args.out_dir);

    // 1) Load URDF -> ManipulatorModel.
    sclerp::urdf::LoadOptions uopt;
    uopt.base_link = args.base_link;
    uopt.tip_link = args.tip_link;
    uopt.collapse_fixed_joints = true;
    uopt.strict = true;

    const auto load = sclerp::urdf::loadManipulatorModelFromFile(args.urdf_path.string(), uopt);
    if (!sclerp::core::ok(load.status)) {
      throw std::runtime_error("loadManipulatorModelFromFile failed: " + load.message);
    }

    // 2) Build kinematics solver.
    const sclerp::core::KinematicsSolver solver(load.model);
    const int dof = solver.model().dof();

    // 3) Choose a safe q_init.
    Eigen::VectorXd q_init = Eigen::VectorXd::Zero(dof);
    if (args.q_init) {
      if (args.q_init->size() != static_cast<std::size_t>(dof)) {
        throw std::runtime_error("--q-init must have exactly " + std::to_string(dof) + " values; got " +
                                 std::to_string(args.q_init->size()));
      }
      for (int i = 0; i < dof; ++i) q_init[i] = args.q_init->at(static_cast<std::size_t>(i));
    } else {
      constexpr double kNearLimitEps = 1e-3;
      for (int i = 0; i < dof; ++i) {
        const auto& lim = solver.model().joint(i).limit;
        if (!lim.enabled) continue;
        if (!std::isfinite(lim.lower) || !std::isfinite(lim.upper)) continue;
        if (!(lim.lower <= lim.upper)) continue;

        const double q0 = 0.0;
        const bool outside = (q0 < lim.lower) || (q0 > lim.upper);
        const bool near_bound = (std::abs(q0 - lim.lower) < kNearLimitEps) ||
                                (std::abs(lim.upper - q0) < kNearLimitEps);
        if (outside || near_bound) {
          q_init[i] = 0.5 * (lim.lower + lim.upper);
        }
      }
    }

    // 4) Build planning request (start pose -> translated goal pose).
    sclerp::core::MotionPlanRequest req;
    req.q_init = q_init;

    const auto st_fk = solver.forwardKinematics(req.q_init, &req.g_i);
    if (!sclerp::core::ok(st_fk)) throw std::runtime_error("forwardKinematics failed");

    req.g_f = req.g_i;
    req.g_f.translation() += Eigen::Vector3d(args.dx, args.dy, args.dz);

    // 5) Create hardcoded obstacles (from local setup).
    std::vector<std::shared_ptr<sclerp::collision::FclObject>> obstacles;
    {
      const sclerp::core::Vec3 dims(1.0, 0.06, 0.29326);
      const sclerp::core::Vec3 pos(-0.2813, -0.5606, 0.14663);
      const double yaw = -std::acos(-1.0) / 4.0;  // -45 deg
      const sclerp::core::Mat3 R = Eigen::AngleAxisd(yaw, Eigen::Vector3d::UnitZ()).toRotationMatrix();

      std::shared_ptr<sclerp::collision::FclObject> box;
      const auto st = sclerp::collision::createBox(dims, pos, R, &box);
      if (!sclerp::core::ok(st)) throw std::runtime_error("collision::createBox failed");
      box->computeAABB();
      obstacles.push_back(std::move(box));
    }

    // 6) Build robot collision meshes from per-link STLs.
    std::vector<std::shared_ptr<sclerp::collision::FclObject>> link_meshes;
    std::vector<sclerp::core::Mat4> mesh_offset_transforms;
    {
      const sclerp::collision::RobotLinkMeshSpec mesh_spec{load.fk_frame_names, load.fk_frame_mesh_uris};
      const auto st = sclerp::collision::buildRobotLinkMeshesFromStlDirectory(solver,
                                                                              args.stl_dir,
                                                                              mesh_spec,
                                                                              &link_meshes,
                                                                              &mesh_offset_transforms);
      if (!sclerp::core::ok(st)) throw std::runtime_error("buildRobotLinkMeshesFromStlDirectory failed");
    }

    std::shared_ptr<sclerp::collision::FclObject> grasped_object;  // none

    const sclerp::collision::CollisionContext ctx{link_meshes, obstacles, grasped_object};
    sclerp::collision::CollisionScene scene{ctx, mesh_offset_transforms};

    // 7) Plan with collision avoidance.
    sclerp::collision::CollisionMotionPlanOptions plan_opt;
    plan_opt.query.check_self_collision = false;
    plan_opt.query.num_links_ignore = 1;
    plan_opt.avoidance.safe_dist = 0.01;

    const auto result = sclerp::collision::planMotionSclerpWithCollision(solver, req, scene, plan_opt);
    if (!sclerp::core::ok(result.status)) {
      throw std::runtime_error("planMotionSclerpWithCollision failed");
    }

    const auto& path = result.path;

    // 8) Time-parameterize (TOTG) + export a CSV for Gazebo playback.
    sclerp::trajectory::Limits lim;
    lim.v_max = Eigen::VectorXd::Constant(dof, args.vmax);
    lim.a_max = Eigen::VectorXd::Constant(dof, args.amax);

    sclerp::trajectory::GridTotg::Config cfg;
    cfg.unwrap_angles = true;

    const sclerp::trajectory::GridTotg totg(lim, cfg);
    sclerp::trajectory::PlannedTrajectory traj;
    const auto st_totg = totg.plan(path, args.sample_dt, &traj, sclerp::trajectory::continuousRevoluteMask(solver.model()));
    if (!sclerp::core::ok(st_totg)) throw std::runtime_error("GridTotg::plan failed");

    const std::filesystem::path csv_path = args.out_dir / "trajectory.csv";
    const auto st_csv = sclerp::trajectory::writeTrajectoryCsv(traj, csv_path.string(), sclerp::trajectory::CsvMode::PositionOnly);
    if (!sclerp::core::ok(st_csv)) throw std::runtime_error("writeTrajectoryCsv failed");

    // 9) Export playback world.
    sclerp::gazebo::WorldRegistry reg;
    registerObstaclesOrThrow(&reg, obstacles);

    sclerp::gazebo::WorldExportOptions opt;
    opt.world_name = args.world_name;

    sclerp::gazebo::RobotModelFromUrdf robot;
    robot.urdf_path = args.urdf_path.string();
    robot.base_link = args.base_link;
    robot.tip_link = args.tip_link;
    robot.collapse_fixed_joints = true;
    robot.link_mesh_stl_directory = args.stl_dir.string();
    opt.robot_from_urdf = robot;

    sclerp::gazebo::JointTrajectoryPlayback playback;
    playback.csv_path = csv_path.string();
    playback.loop = args.loop;
    playback.rate = args.rate;
    opt.joint_trajectory = playback;

    if (args.plugin_dir) {
      opt.joint_trajectory_plugin_filename = (args.plugin_dir->empty())
          ? opt.joint_trajectory_plugin_filename
          : (args.plugin_dir.value() / opt.joint_trajectory_plugin_filename).string();
    }

    const std::filesystem::path world_path = args.out_dir / "world.sdf";
    const auto st_world = reg.writeSdfWorld(world_path.string(), opt);
    if (!sclerp::core::ok(st_world)) {
      throw std::runtime_error("WorldRegistry::writeSdfWorld failed");
    }

    std::cout << "Wrote:\n"
              << "  world:      " << world_path << "\n"
              << "  trajectory: " << csv_path << "\n\n";

    std::cout << "Run Gazebo:\n";
    if (!args.plugin_dir) {
      std::cout << "  export IGN_GAZEBO_SYSTEM_PLUGIN_PATH=\"<sclerp_build>/gazebo:${IGN_GAZEBO_SYSTEM_PLUGIN_PATH}\"\n";
    }
    std::cout << "  ign gazebo \"" << world_path << "\"\n";

    return 0;
  } catch (const std::exception& e) {
    std::cerr << "[example_2_hardcoded_obstacles_gazebo] ERROR: " << e.what() << "\n";
    return 1;
  }
}

File: python/scripts/example_2_hardcoded_obstacles_gazebo.py
#!/usr/bin/env python3

from __future__ import annotations

import argparse
import math
import sys
from pathlib import Path

import numpy as np

import sclerp

from example_common import (
    add_common_cli,
    ensure_output_dir,
    maybe_midpoint_q_init,
    parse_common_args_or_throw,
    print_gazebo_hint,
    rpy_to_rot,
)


def main() -> int:
    p = argparse.ArgumentParser(
        description="Example 2 (Python): plan with hardcoded obstacles (collision), TOTG, export Gazebo world + CSV."
    )
    add_common_cli(p)
    ns = p.parse_args()

    args = parse_common_args_or_throw(ns)
    ensure_output_dir(args.out_dir)

    # 1) Load URDF -> ManipulatorModel.
    uopt = sclerp.urdf.LoadOptions()
    uopt.base_link = args.base_link
    uopt.tip_link = args.tip_link
    uopt.collapse_fixed_joints = True
    uopt.strict = True

    load = sclerp.urdf.load_manipulator_model_from_file(str(args.urdf_path), uopt)
    if not sclerp.core.ok(load.status):
        raise RuntimeError(f"load_manipulator_model_from_file failed: {load.message}")

    # 2) Solver.
    solver = sclerp.core.KinematicsSolver(load.model)
    dof = solver.model().dof()

    # 3) q_init.
    if args.q_init is None:
        q_init = maybe_midpoint_q_init(solver, np.zeros((dof,), dtype=np.float64))
    else:
        q_init = np.asarray(args.q_init, dtype=np.float64).reshape(-1)
        if q_init.shape != (dof,):
            raise RuntimeError(f"--q-init must have exactly {dof} values; got {q_init.size}")

    # 4) Start pose -> translated goal.
    st_fk, g_i = solver.forward_kinematics(q_init)
    if not sclerp.core.ok(st_fk):
        raise RuntimeError("forward_kinematics failed")
    g_f = np.array(g_i, copy=True)
    g_f[:3, 3] += np.array([args.dx, args.dy, args.dz], dtype=np.float64)

    # 5) Hardcoded obstacle box.
    dims = np.array([1.0, 0.06, 0.29326], dtype=np.float64)
    pos = np.array([-0.2813, -0.5606, 0.14663], dtype=np.float64)
    yaw = -math.pi / 4.0
    R = rpy_to_rot(0.0, 0.0, yaw)

    st_box, box = sclerp.collision.create_box(dims, pos, R)
    if not sclerp.core.ok(st_box):
        raise RuntimeError(f"create_box failed: status={st_box}")
    box.compute_aabb()
    obstacles = [box]

    # 6) Robot link meshes.
    spec = sclerp.collision.RobotLinkMeshSpec(load.fk_frame_names, load.fk_frame_mesh_uris)
    st_mesh, link_meshes, mesh_offsets = sclerp.collision.build_robot_link_meshes_from_stl_directory(
        solver, str(args.stl_dir), spec
    )
    if not sclerp.core.ok(st_mesh):
        raise RuntimeError(f"build_robot_link_meshes_from_stl_directory failed: status={st_mesh}")

    # 7) Plan with collision.
    plan_opt = sclerp.collision.CollisionMotionPlanOptions()
    plan_opt.query.check_self_collision = False
    plan_opt.query.num_links_ignore = 1
    plan_opt.avoidance.safe_dist = 0.01

    if args.rmrc_nullspace:
        plan_opt.motion.rmrc.nullspace.enabled = True
        plan_opt.motion.rmrc.nullspace.joint_limits.enabled = True
        plan_opt.motion.rmrc.nullspace.posture.enabled = True

    result = sclerp.collision.plan_motion_sclerp_with_collision(
        solver, q_init, g_f, link_meshes, obstacles, mesh_offsets, plan_opt
    )
    if not sclerp.core.ok(result.status):
        raise RuntimeError(f"plan_motion_sclerp_with_collision failed: status={result.status}")

    # 8) TOTG.
    lim = sclerp.trajectory.Limits()
    lim.v_max = np.full((dof,), args.vmax, dtype=np.float64)
    lim.a_max = np.full((dof,), args.amax, dtype=np.float64)

    cfg = sclerp.trajectory.GridTotg.Config()
    cfg.unwrap_angles = True
    totg = sclerp.trajectory.GridTotg(lim, cfg)
    unwrap_mask = sclerp.trajectory.continuous_revolute_mask(solver.model())

    st_totg, traj = totg.plan(result.path, args.sample_dt, unwrap_mask)
    if not sclerp.core.ok(st_totg):
        raise RuntimeError(f"GridTotg.plan failed: status={st_totg}")

    csv_path = args.out_dir / "trajectory.csv"
    st_csv = sclerp.trajectory.write_trajectory_csv(traj, str(csv_path), sclerp.trajectory.CsvMode.PositionOnly)
    if not sclerp.core.ok(st_csv):
        raise RuntimeError(f"write_trajectory_csv failed: status={st_csv}")

    # 9) Export world (robot + obstacles + playback).
    reg = sclerp.gazebo.WorldRegistry()
    st_reg = reg.register_obstacle(box, "obstacle_box")
    if not sclerp.core.ok(st_reg):
        raise RuntimeError(f"register_obstacle failed: status={st_reg}")

    wopt = sclerp.gazebo.WorldExportOptions()
    wopt.world_name = args.world_name

    robot = sclerp.gazebo.RobotModelFromUrdf()
    robot.urdf_path = str(args.urdf_path)
    robot.base_link = args.base_link
    robot.tip_link = args.tip_link
    robot.collapse_fixed_joints = True
    robot.link_mesh_stl_directory = str(args.stl_dir)
    wopt.robot_from_urdf = robot

    playback = sclerp.gazebo.JointTrajectoryPlayback()
    playback.csv_path = str(csv_path)
    playback.loop = args.loop
    playback.rate = args.rate
    wopt.joint_trajectory = playback

    if args.plugin_dir and str(args.plugin_dir):
        wopt.joint_trajectory_plugin_filename = str(Path(args.plugin_dir) / wopt.joint_trajectory_plugin_filename)

    world_path = args.out_dir / "world.sdf"
    st_world = reg.write_sdf_world(str(world_path), wopt)
    if not sclerp.core.ok(st_world):
        raise RuntimeError(f"write_sdf_world failed: status={st_world}")

    print("Wrote:")
    print(f"  world:      {world_path}")
    print(f"  trajectory: {csv_path}")

    print_gazebo_hint(world_path)
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as e:
        print(f"[example_2_hardcoded_obstacles_gazebo.py] ERROR: {e}", file=sys.stderr)
        raise

